Improve rendering Markdown lists in a terminal

fingolfin opened the pull request:
This changes how Markdown lists are rendered in the Terminal:
- the list is no longer indented by two spaces
- the bullet points are now followed by a single space, not two
- extra lines in an entry are indented correctly across all levels
- fix printing / indentation for ordered lists starting at an index other than 1: the padding for the labels was being computed incorrectly
- fix word wrap with a given column limit

Consider this Markdown input:
```md
Before the list:
- top level\
  with an extra line
  - second level\
    again with an extra line
    - third level\
      yet again with an extra line
      - fourth level\
        and another extra line
        - fifth level\
          final extra line
- back to top level
```

Then with Julia master it renders in the terminal like this:
```
  Before the list:

    •  top level
       with an extra line
       –  second level
        again with an extra line
          ▪  third level
         yet again with an extra line
             –  fourth level
          and another extra line
                ▪  fifth level
           final extra line
    •  back to top level
```

With this PR:
```
  Before the list:

  • top level
    with an extra line
    – second level
      again with an extra line
      ▪ third level
        yet again with an extra line
        – fourth level
          and another extra line
          ▪ fifth level
            final extra line
  • back to top level
```

Resolves #40508, resolves #58514

fingolfin committed:
Improve rendering of Markdown lists in terminal

This changes how Markdown lists are rendered in the Terminal:
- the list is no longer indented by two spaces
- the bullet points are now followed by a single space, not two
- extra lines in an entry are indented correctly across all levels

Consider this Markdown input:
```md
Before the list:
- top level\
  with an extra line
  - second level\
    again with an extra line
    - third level\
      yet again with an extra line
      - fourth level\
        and another extra line
        - fifth level\
          final extra line
- back to top level
```

Then with Julia master it renders in the terminal like this:
```
  Before the list:

    •  top level
       with an extra line
       –  second level
        again with an extra line
          ▪  third level
         yet again with an extra line
             –  fourth level
          and another extra line
                ▪  fifth level
           final extra line
    •  back to top level
```

With this PR:
```
  Before the list:

  • top level
    with an extra line
    – second level
      again with an extra line
      ▪ third level
        yet again with an extra line
        – fourth level
          and another extra line
          ▪ fifth level
            final extra line
  • back to top level
```

fingolfin committed:
tweak

fingolfin committed:
Also handle ordered lists

tecosaur commented:
I haven't looked at these changes in detail, but from the description these changes look good.

tecosaur reviewed (approved):
Yep, the code changes look sensible. I don't see any reason why this shouldn't be merged.

fingolfin committed:
Adjust a test to new indentation

fingolfin commented:
OK I am fixing the tests now. This already revealed another hidden bug in the code (word wrapped uses a hardcoded constant 10 that really ought to depend on the margin & label width)

And another: bullet padding assumed all ordered lists start at 1... *sigh*

Oh man, this is growing fast... each time I fix one bug I find another 😅

fingolfin committed:
Tighten tests

fingolfin committed:
Markdown: fix label width in ordered lists

Markdown lists can start at an arbitrary index, not just 1.

fingolfin committed:
Fix word wrapping in Markdown lists

Consider this input:
```julia
using Markdown
doc =
    md"""
    1. a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij

       - a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij

         - a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij
    """;
println(sprint(Markdown.term, doc, 50))
```

With Julia 1.12, it prints:
```md
    1. a bc def ghij a bc def ghij a bc def
       ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def
       ghij a bc def ghij

       •  a bc def ghij a bc def
       ghij a bc def ghij a bc
       def ghij a bc def ghij a
       bc def ghij a bc def ghij
       a bc def ghij a bc def ghij

       •  a bc def ghij a
       bc def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij
```

With Julia `master` (note the regression (?) in that an empty line is missing...):
```md
    1. a bc def ghij a bc def ghij a bc def
       ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def
       ghij a bc def ghij

       –  a bc def ghij a bc def
        ghij a bc def ghij a bc
        def ghij a bc def ghij a
        bc def ghij a bc def ghij
        a bc def ghij a bc def ghij
          ▪  a bc def ghij a
         bc def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij
```

With my other Markdown list tweaks:
```md
  1. a bc def ghij a bc def ghij a bc def
     ghij a bc def ghij a bc def ghij a
     bc def ghij a bc def ghij a bc def
     ghij a bc def ghij

     – a bc def ghij a bc def
       ghij a bc def ghij a bc
       def ghij a bc def ghij a
       bc def ghij a bc def ghij
       a bc def ghij a bc def ghij
       ▪ a bc def ghij a
         bc def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij
```

After this PR:
```md
  1. a bc def ghij a bc def ghij a bc def ghij a
     bc def ghij a bc def ghij a bc def ghij a bc
     def ghij a bc def ghij a bc def ghij

     – a bc def ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def ghij
       ▪ a bc def ghij a bc def ghij a bc def ghij
         a bc def ghij a bc def ghij a bc def ghij
         a bc def ghij a bc def ghij a bc def ghij
```

fingolfin commented:
OK I realized that rendering an ordered list starting at e.g. index 999 where the padding for the list labels was computed incorrectly.

Before this PR:
```md
julia> md"""
        999. a
        1000. b
       """
    999. a
    1000. b
```
After this PR:
```md
julia> md"""
        999. a
        1000. b
       """
   999. a
  1000. b
```


Moreover, this PR is now fixing word wrapping in Markdown lists.

To see the issue, consider this input:
```julia
using Markdown
doc =
    md"""
    1. a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij

       - a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij

         - a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij a bc def ghij
    """;
println(sprint(Markdown.term, doc, 50))
```

With Julia 1.12, it prints:
```md
    1. a bc def ghij a bc def ghij a bc def
       ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def
       ghij a bc def ghij

       •  a bc def ghij a bc def
       ghij a bc def ghij a bc
       def ghij a bc def ghij a
       bc def ghij a bc def ghij
       a bc def ghij a bc def ghij

       •  a bc def ghij a
       bc def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij a bc
       def ghij
```

With Julia `master` (note the regression (?) in that an empty line is missing... I have not yet looked into that...):
```md
    1. a bc def ghij a bc def ghij a bc def
       ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def
       ghij a bc def ghij

       –  a bc def ghij a bc def
        ghij a bc def ghij a bc
        def ghij a bc def ghij a
        bc def ghij a bc def ghij
        a bc def ghij a bc def ghij
          ▪  a bc def ghij a
         bc def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij
```

With my other Markdown list tweaks:
```md
  1. a bc def ghij a bc def ghij a bc def
     ghij a bc def ghij a bc def ghij a
     bc def ghij a bc def ghij a bc def
     ghij a bc def ghij

     – a bc def ghij a bc def
       ghij a bc def ghij a bc
       def ghij a bc def ghij a
       bc def ghij a bc def ghij
       a bc def ghij a bc def ghij
       ▪ a bc def ghij a
         bc def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij a bc
         def ghij
```

After this PR:
```md
  1. a bc def ghij a bc def ghij a bc def ghij a
     bc def ghij a bc def ghij a bc def ghij a bc
     def ghij a bc def ghij a bc def ghij

     – a bc def ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def ghij a
       bc def ghij a bc def ghij a bc def ghij
       ▪ a bc def ghij a bc def ghij a bc def ghij
         a bc def ghij a bc def ghij a bc def ghij
         a bc def ghij a bc def ghij a bc def ghij
```

fingolfin commented:
Ah, the "missing" line is due to @tecosaur's PR ##55456. I am fine with that change, actually, but what I don't get is why it is applied unevenly.

tecosaur commented:
> Ah, the "missing" line is due to tecosaur's PR #https://github.com/JuliaLang/julia/pull/55456. I am fine with that change, actually, but what I don't get is why it is applied unevenly.

I suspect this may be due to how the tight/loose attribute of markdown lists is set. I have a vague recollection of this.

fingolfin commented:
Yeah the tight/loose handling is also buggy. But anyway, that's a separate issue.
